compile using babel for compatibility with older versions

brettz9 opened the pull request:
This should work (to close #95), but still need to add matrix logic to travis to install older nyc, etc. to confirm build is working

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,35 @@
-  - npm run build
   - npm run test:ci
 node_js:
+  - 0.10
```
I think you need to quote this or risk it being interpreted as "0.1" because... yaml

brettz9 replied:
Seems to have been interpreted ok by Travis as is.

brettz9 replied:
I can of course quote it though if you prefer.

michaelficarra replied:
As long as it works. I recall something like this not working in the past.

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,35 @@
+before_install:
+  - npm config set depth 0
+before_script: >
+  node_version=$(node -v);
```
You can get the major directly with `node -v | cut -d . -f 1 | tail -c +2`

brettz9 replied:
If testing against 0.12, IIRC, the 0 isn't shown.

brettz9 commented:
The latest Rollup plugins don't work on older Node. I've avoided the build step for those versions now, but now tests are failing as Babel is insufficient as Babel doesn't polyfill globals like `Object.entries` by default. Looking to add plugin to transform now.

brettz9 commented:
It seems the Babel plugin for dot-all (the s flag) will only handle transforming when within regex literals, so one test is failing on Node 6. I could perhaps add core-js if it has a `RegExp` polyfill that can handle the flag.

As far as earlier Node versions, `@babel/register` itself does not work on older Node. I can look to using `babel-register`, but then the dev. code will be using a less up-to-date tool. (I could perhaps see about running a different test script on Travis.)

brettz9 commented:
So `babel-register` doesn't work as it doesn't seem to be able to understand the `import` statements as `@babel/register` does.

I could try retooling the tests to point to the `dist` copy if that is desired.

Alternatively, we could just avoid adding pre-Node 6 versions to travis and get out a patch earlier, trusting that babel is working as advertised.

As far as the single Node 6 test failure, from what I can tell, the core-js polyfill for the RegExp constructor doesn't properly handle the dotall flag as I was still getting errors despite including the polyfill and successfully confirming it was being used. But older esquery dependencies shouldn't have been relying on the 's' flag anyways, as it had not been supported in esquery previously (nor would older environments support it), so while it would be good to see what we can do to get it to work under core-js, I think we could also drop Node 6 from Travis for now too (or ignore its result), given that this one failure is expected without a proper polyfill.

Suggested steps forward?

brettz9 commented:
Btw:

1. For `engines`, I set a minimum of 0.10 based on [this comment](https://github.com/estools/esquery/issues/95#issuecomment-605348012), though let me know if you want to test against 0.1.14, as this is, IIRC, the earliest usable).
2. FWIW, for `rollup-plugin-babel`, I didn't exclude `node_modules` from babelification in case the latest `estraverse` is using anything incompatible.

mbroadst commented:
@brettz9 what if we use a modern version of node to build everything, and then only run the tests in legacy versions of node? I was able to run the `esquery` tests just fine in Node 4 with the work present in your initial PR, so we might be solving for the wrong thing at this point.

There are a few ways I could imagine doing this:
- I think travis already uses `nvm`, so you might be able to use that command before running the actual tests
- We could get relatively complex and use [build stages](https://docs.travis-ci.com/user/build-stages) and an S3 bucket, where the built assets are fetched from the S3 bucket before each build.

brettz9 commented:
@mbroadst : Since @michaelficarra had indicated previously that he did not wish to bundle the distributions, yes, I think another way would need to be adopted to build (or obtain) the builds.

Even if the building issues are addressed, those earlier Node environments will still fail on the `dotAll` test, though as mentioned, and as a new feature anyways, I think that's one that could be ignored for now.

(Btw, I don't see though how the original PR could have worked for you on Node 4 with the dotAll test nor. I also don't see how it could not have erred due to a lack of `Object.entries` unless you were using your own polyfill like core-js. The `Object.entries` change was especially important as it impacted the build and is not merely testing related.)

This is toward the end of my day over here (in China), btw, so you might want to look into devising a build solution on your own if you need a fix for the day.

mbroadst commented:
@brettz9 I'm looking into this now.

You're correct that I hadn't got to the `Object.entries` issues yet, I was just confirming I could load the emitted javascript, thanks for that tip!

I think we still have more work to do here outside of the `dotAll` test for older distributions. All of the test files are currently using ES6 modules (and likely other newer features), so they will have be transpiled as well in order to run in those environments. Involving so much tooling and transpiling is appearing to have diminishing returns in terms of future maintenance of this codebase.

brettz9 commented:
By moving toward ESM, I would think it would be better preparing for future maintenance (and is already doing so in allowing for ESM distributions). But yes, it is causing pains at the moment due in part to the fact that the toolers have themselves dropped support.

How about I adapt the tests to avoid ESM in favor of pointing to the compiled `dist` files (and skip the dotAll test). If @michaelficarra wants the tests in ESM, we can add back the ESM tests in a future breaking version (which drops testing for older versions). But this would allow tests to work for now, while still allowing ESM source and builds.

mbroadst commented:
@brettz9 I'm not against moving towards ESM at all, I'm just pondering the changes needed in order to support what's needed here compared to simply cutting a major release now. I think the approach you suggest would work, but we still end up with a travis configuration which is manually replacing dependencies for certain versions of node (etc) which is more what I'm getting at in terms of a maintenance burden.

michaelficarra commented:
@brettz9 I am okay with limiting the new functionality to only the supported engines going forward. This means that the library just needs to load and provide at least the 1.1.x functionality on older platforms. Will that help simplify this PR? We can gate tests for newer features behind node version checks.

brettz9 commented:
> @brettz9 I am okay with limiting the new functionality to only the supported engines going forward. This means that the library just needs to load and provide at least the 1.1.x functionality on older platforms. Will that help simplify this PR? We can gate tests for newer features behind node version checks.

While gating will help us avoid breaking for older Node (with the `dotAll` feature) (and I've just added such a commit in this PR), this will only work as far back as Node 6.

The bigger problems are:
1. Unless we change the tests to use `require` statements against the `dist` UMD builds, we need a more modern version of `@babel/register` that can understand `import`.
2. If you want the latest testing tools--e.g., nyc (including for the fact that they may patch dev. vulnerabilities), these have increased minimum versions of Node. And even if you don't care about this, it would involve some work finding out which versions will still work.

Based on this and the desire to provide a patch to avoid breakage, my personal suggestion would be:

1. Accept a separate PR (which I can prepare) and then publish a new patch version which goes back to having known compatibility, reverting the recent additions entirely (I think this would be safer, even while it is possible that this PR could possibly avoid breakage.)
2. Subsequently, accept this PR as a breaking change (on top of the just-mentioned PR), with a minimum of Node 6 (and added to `engines` as such) going forward.

This way, we can be sure older users have a migration path which is known not to break, and those using Node >= 6 can be quickly informed in the release notes that the only breaking change is Node compatibility (I would expect at least eslint could bump in such a case, especially as their Node minimum is higher than that). Though there is some branching needed in Travis to support Node 6, it would be much less complicated than trying to dig through very old versions of nyc, etc.

mbroadst commented:
@michaelficarra gentle nudge to see if there has been any decision made here?

michaelficarra commented:
@mbroadst I never mind a ping. Sorry about the delay, the week of a TC39 meeting is always nuts.

@brettz9 I don't think we need to go through excessive effort to get old versions of node running our tests in CI. What we really care about is that the code we actually distribute will work in these environments. If you've manually tested it by converting the imports to requires and running the tests, that's sufficient. Is that enough guidance for you to move forward?

brettz9 commented:
Ok, after reverting tests from using imports (and pointing to `dist`), reverting other ES6 features in test source, and using Mocha 3 (though without attempting nyc checks) and dropping babel/register for older Node support (FWIW, this is all in my `test-old-node-dist` branch), I've now confirmed that the mocha tests are passing on old versions of Node as far back as 0.10.0.

I was not able to successfully install earlier versions of Node (using nvm) to find out whether earlier versions might work.

So it seems pretty clear that his will continue to work on versions as far back as 0.10.0.

@mbroadst , do you have any objections to my keeping the addition of this PR to add `engines` with a minimum Node version of 0.10.0? It may technically be breaking, but since we can't verify it ever worked before 0.10.0, I'm not sure this should be such a big concern (and might be considered more of a "fix" since we are specifying what has been the expected behavior--that we only know it to work with 0.10.0+).

Btw, I've added a commit to drop testing previous to Node 6 (since it seems with the manual testing I've now done, you would like to keep the modern testing apparatus for the next (patch) release). And you can see that all automated tests are now passing as a result.

brettz9 commented:
One minor problem I have not been able to work around--though not sure we need to since `npm test` doesn't fail despite the 100% threshold--is that there is an apparent nyc bug (or `@babel/register`) whereby this PR causes nyc to report one line as not having an `else` condition covered, but the line in question (113 of esquery.js) is indeed being covered and has no `else`. Since this is not even a red `if` portion that is being reported, I think this very minor coverage regression is safe to ignore.

mbroadst commented:
@brettz9 I think that's probably fine, thank you very much for your hard work!

brettz9 commented:
@michaelficarra In case you didn't get a notice, this should now be ready for review.

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,45 @@
+        if [ ${node_version:1:1} -ge 6 ]; then
+          echo "Node 6+"
+          npm install --no-save "eslint@5" mocha@6.2.2 nyc@14.1.1
+          npm run test:ci
```
How are we running the tests without building (`npm run build`) on this platform?

michaelficarra replied:
Oh I see now. By not running the build step, we're testing against the old parser and will fail when we go to add a new feature.

brettz9 replied:
I don't recall if the build routine works or not (I can try for parity if you like), but it is not needed because the tests are able to run using `@babel/register` testing against the source instead of `dist` files.

(I only needed to switch to testing the `dist` files in the branch where I manually confirmed the old Node builds were still working. But that branch had to undo ESM in the test files and fixtures, avoid other ES6 features, as well as revert to the vulnerable Mocha 3.)

michaelficarra replied:
I don't understand. Why would we be able to run "against the source" using `@babel/register` but not compile? Isn't `@babel/register` just compiling on the fly and running that?

brettz9 replied:
On the much older Node versions, `@babel/register` doesn't work, and `babel-register`  doesn't seem to be able to understand the import statements as @babel/register does.

As far as Node 6 that is failing on Travis, yes, `@babel/register` is compiling on the fly so the babel component should be doable, but it appears that the current `@rollup/pluginutils` (used by the current versions of `@rollup/plugin-commonjs` and `@rollup/plugin-node-resolve` needed in the Rollup config file) is using object spread properties which weren't supported in Node 6. Note that it wouldn't be enough to go back just to an older version here either, as these were previously not in the `@rollup` scope but were their own packages. But if we go too far back in our toolchain, then certain other features might not have been supported then.

Is it really critical that the build scripts are confirmed to work? These aren't available to work in the npm releases anyways (since the rollup config is not in `package.json` `files` and Rollup, etc. are devDeps).

This should really only be needed for those compiling from source on the repo who would presumably have access to more up to date machines. Seems it may be some unnecessary work for little benefit (not to mention adding to the travis build time).

But I can try to see if we can get it to work for Node 6 if it is important.

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,45 @@
+          npm install --no-save "eslint@5" mocha@6.2.2 nyc@14.1.1
+          npm run test:ci
+        else
+          if [ ${node_version:1:1} -ge 4 ]; then
```
This looks like dead code since we only list node 6, 8, 10, 12 in the config above.

brettz9 replied:
Gotcha, fixed (I can't confirm Travis builds are ok now as the site is not loading for me here).

michaelficarra commented on tests/queryAttribute.js:
```diff
@@ -177,12 +177,14 @@ describe('Attribute query', function () {
-            literal.body[0].declarations[0].init
-        ]);
-    });
+    if ((/^(?:[89]|\d{2,})\./).test(process.version)) {
```
Would this be a little more readable?

```suggestion
    if (parseInt(process.version) >= 8) {
```

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,33 @@
 
+before_install:
+  - npm config set depth 0
+before_script: >
```
Would it be easier to use the `matrix` feature that we use below for switching on node version instead of this?

michaelficarra commented on .travis.yml:
```diff
@@ -1,13 +1,33 @@
   - 12
 
+before_install:
+  - npm config set depth 0
```
What does this do for us?

brettz9 replied:
Removed.

michaelficarra commented on .travis.yml:
```diff
@@ -4,10 +4,28 @@ script:
+  node_version=$(node -v);
+  if [ ${node_version:3:1} = "." ]; then
+    echo "Node 10+"
+    npm install --no-save "eslint@6"
```
We don't actually need this line, right? It should already be installed.

brettz9 replied:
I think we need the `npm install`, but the rest should be omittable, yes. Removed.

michaelficarra commented:
@brettz9 Now that I enabled it again, I see that `npm run build` fails on node 6.  Can you look into it?

michaelficarra commented on .travis.yml:
```diff
@@ -13,7 +13,7 @@ before_script: >
   if [ ${node_version:3:1} = "." ]; then
     echo "Node 10+"
-    npm install --no-save "eslint@6"
+    npm install
```
Won't this just case `npm install` to happen twice on node 10+? It is already run by travis by default.

brettz9 replied:
Ah, ok, I've added a commit to remove.

michaelficarra commented:
This is good enough for now. We'll deal with node 6 CI issues if/when they come up in the future. Thanks for the hard work, @brettz9.

brettz9 commented:
Thanks for all the thoughtful reviewing and merges!
